### PR TITLE
refactor: single binary

### PR DESCRIPTION
**Pull request by decentral1se**
See https://github.com/cblgh/cerca/issues/74.

Probably also provoking a violent merge conflict but I had a burst of energy 🙃 Can hold off and rebase over https://github.com/cblgh/cerca/pull/92, no stress. ~~Just wanted to get some early feedback on this one.~~ **update**: [seems good](https://github.com/cblgh/cerca/pull/93#issuecomment-2497162117) 

I've looked at each subcommand and realised that I'd probably like to have it all under the one `cerca` binary. This is from my perspective the most convenient setup for making maintenance easy and straightforward.

I've merged the various subcommand files in under `./cmd/cerca/...`. Following the example [here](https://gobyexample.com/command-line-subcommands), you can now do e.g. `cerca [migrate|admin|user|reset] -h`.

The diff is a bit wild due to filename renames but the main things I've done are 1) refactor to use a single `inform`/`complain` 2) changed the `main()` of each subcommand file to the filename, i.e. `migrate()`, `admin()`, etc. 3) put a `main()` in `cerca.go` to route the parsing.

~~I'd probably merge `user`/`reset` into the same subcommand? I can draw up a design for the CLI surface if you are up for it. The rest is just cleaning it up and doing a bunch of manual testing.~~ **update**: done

### CODE REVIEW

**Comment by cblgh** [comment]
ahhh sick! will have a look at the code soon but wanted to say i am a big supporter of this change. thanks for championing it!On 24 Nov 2024, at 18:45, decentral1se ***@***.***> wrote:﻿See #74.
Probably also provoking a violent merge conflict but I had a burst of energy 🙃 Can hold off and rebase, no stress. Just wanted to get some early feedback on this one.
I've looked at each subcommand and realised that I'd probably like to have it all under the one cerca binary. This is from my perspective the most convenient setup for making maintenance easy and straightforward.
I've merged the various subcommand files in under ./cmd/cerca/.... Following the example here, you can now do e.g. cerca [migrate|admin|user|reset] -h.
The diff is a bit wild due to filename renames but the main things I've done are 1) refactor to use a single inform/complain 2) changed the main() of each subcommand file to the filename, i.e. migrate(), admin(), etc. 3) put a main() in cerca.go to route the parsing.
I'd probably merge user/reset into the same subcommand? I can draw up a design for the CLI surface if you are up for it.



You can view, comment on, or merge this pull request online at:
  https://github.com/cblgh/cerca/pull/93

Commit Summary

  2e5614c  chore: remove unused default about.md
  3edf849  refactor: move cerca binary under cmd/
  f7531a5  WIP: merge cmds into single cmd


File Changes  (7 files)

  
    M
    .gitignore
    (1)
  
  
    D
    about.md
    (51)
  
  
    D
    cmd/add-admin/main.go
    (75)
  
  
    D
    cmd/admin-add-user/main.go
    (103)
  
  
    D
    cmd/admin-reset/main.go
    (74)
  
  
    D
    cmd/migration-tool/main.go
    (71)
  
  
    D
    run.go
    (76)
  


Patch Links:

  https://github.com/cblgh/cerca/pull/93.patch
  https://github.com/cblgh/cerca/pull/93.diff

**Comment by decentral1se** [comment]
> ahhh sick! will have a look at the code soon but wanted to say i am a big supporter of this change. thanks for championing it!

@cblgh oh cool, great! sure lemme get something cleaned up and reviewable and then I'll ping for the review labours 🧠 thankksss

**Comment by decentral1se** [comment]
Aight this is ready for review 🎉

```
NAME:
  cerca - lean forum software

USAGE:
  cerca [command] [options]

COMMANDS:
  admin     promote user to admin
  migrate   manage database migrations
  user      create new user
  reset     reset user password

GLOBAL OPTIONS:
  -help
        show help (default: false)

OPTIONS:
  -allowlist string
    	domains which can be used to read verification codes from during registration
  -authkey string
    	session cookies authentication key
  -config string
    	config and settings file containing cerca's customizations (default "cerca.toml")
  -data string
    	directory where cerca will dump its database (default "./data")
  -dev
    	trigger development mode
```

**Comment by cblgh** [comment]
@decentral1se niiiiiiiice! i will tend to it this week :> down with a cold unfortunately and recently home from a long rail travel so will be a bit slower than last week

File: cmd/cerca/cerca.go [hunk 1]
```diff
@@ -27,24 +27,63 @@ func readAllowlist(location string) []string {
...
+        show help (default: false)
+
+OPTIONS:
+`
+	usage := func() {
+		fmt.Fprintf(os.Stderr, help)
+		flag.PrintDefaults()
+	}
+	flag.Usage = usage
+
```
Comment (cblgh) [inline cmd/cerca/cerca.go:84]: a new section in the readme that dumps the commands usage info you've written here / is output when running cerca --help would also be a good addition i think! when the readme was originally authored, there was pretty much just the one command lol :)

File: about.md
Comment (cblgh) [inline about.md:1]: generally not a fan of these kind of unrelated changes in PRs, but in this case i think it was overdue! thanks for catching it :)

File: cmd/cerca/cerca.go [hunk 2]
```diff
@@ -27,24 +27,63 @@ func readAllowlist(location string) []string {
...
 	var configPath string
 	var dataDir string
 	var dev bool
+
 	flag.BoolVar(&dev, "dev", false, "trigger development mode")
 	flag.StringVar(&allowlistLocation, "allowlist", "", "domains which can be used to read verification codes from during registration")
 	flag.StringVar(&sessionKey, "authkey", "", "session cookies authentication key")
 	flag.StringVar(&configPath, "config", "cerca.toml", "config and settings file containing cerca's customizations")
 	flag.StringVar(&dataDir, "data", "./data", "directory where cerca will dump its database")
+
```
Comment (cblgh) [inline cmd/cerca/cerca.go:60]: can we drop the first-line `NAME:` from each help output?

Comment (cblgh) [reply cmd/cerca/cerca.go:60]: thinking a bit more, maybe a util function in `cmd/cerca/` would be a good addition and reduce string repetition? 

i also think it is important to not lose concrete example invocations for each commands in this refactor. contrast the following two for the migration tool:

```
usage
	migration-tool --list 
	migration-tool --migration <name of migration>
```

is a lot more useful for orienting as a confused user / infrequent maintainer than

```
  cerca migrate [command] [options]
```

---
help util sketch

```golang
func createHelpString(commandName, usageOptions, exampleUsage string) string {
  return fmt.Sprintf(`%s
  USAGE:
  \t%s
  \t%s
  GLOBAL OPTIONS:
  \t-help
  \t\tshow help
  OPTIONS:`, commandName, usageOptions, exampleUsage)
}
```

`exampleUsage` could cover the expanded `COMMANDS: ` section for the main command

Comment (decentral1se) [reply cmd/cerca/cerca.go:60]: @cblgh nice yep, can do. what do you mean by `usageOptions` in this example? the options are below in `OPTIONS: ...`? do you mean a `cerca [command] [options]` thing or 2 concrete examples?

Comment (cblgh) [reply cmd/cerca/cerca.go:60]: @decentral1se idk lol i think i initially was just putting it in there as a placeholder for your `cerca [command] [options]` string, with `exampleUsage` representing the examples proper. 

could totally skip the `[command] [options]` line though, and simply stick to outputting the name, the concrete examples (or in the case of the main `cerca` cmd, the `COMMANDS` listing), and finishing with section `OPTIONS` with `printDefaults` like you have atm

wdyt?

Comment (decentral1se) [reply cmd/cerca/cerca.go:90]: looks a bit out of place to include the name when we just ran the command and the name is in the terminal. the rest sounds good! so here's what i came up with: https://github.com/cblgh/cerca/pull/93/commits/7151cb5e910cb885534d8da5b7c918ed6c95202c

File: cmd/cerca/cerca.go (hunk 1 above)
Comment (decentral1se) [reply cmd/cerca/cerca.go]: @cblgh https://github.com/cblgh/cerca/pull/93/commits/da4f19c715eeb773a0d333d7b728cfa3e40af341 and [rendered](https://github.com/cblgh/cerca/blob/da4f19c715eeb773a0d333d7b728cfa3e40af341/README.md#usage)

File: cmd/cerca/cerca.go (hunk 2 above)
Comment (cblgh) [reply cmd/cerca/cerca.go:90]: oh i like it!

> looks a bit out of place to include the name on a newline when we just ran the command and the name is in the terminal

i think i can live with this injury to my aesthetic sensibilities X)

thanks!

File: cmd/cerca/cerca.go [hunk 3]
```diff
@@ -11,6 +11,36 @@ import (
...
 
+func createHelpString(usageExamples []string, isMainCmd bool) string {
+	helpString := fmt.Sprintf("USAGE:\n    %s\n", strings.Join(usageExamples, "\n    "))
+
+	if isMainCmd {
+		helpString += "\nCOMMANDS:" + `
+    admin     promote user to admin
+    migrate   manage database migrations
+    user      create new user
+    reset     reset user password` + "\n"
```
Comment (cblgh) [inline cmd/cerca/cerca.go:22]: in the shower (of course, where else) i thought that maybe the naming of the commands could be improved.

`cerca admin` doesn't tell enough about its function, while `cerca make-admin` does. following that line of thought, what do you think about the following command names? and to also reorder them based on alphabetic order, also below

1. user -> adduser 
2. admin -> makeadmin
3. migrate -> migrate
4. reset -> resetpw

Comment (cblgh) [reply cmd/cerca/cerca.go:22]: actually fuk it i'm just going to merge and do some last minute tidying up as a way to tie the bow on this super refactor! thank you d1 :muscle: :heart:

**Pull request merged**